### PR TITLE
Split and Concat operations

### PR DESCRIPTION
**Pull request by mdanilow**
Testbenches, top functions and test.tcl scripts included and tested.
The operations are also integrated and tested in the finn compiler fork.

### CODE REVIEW

File: concat.hpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
+#include "utils.hpp"
```
Comment (preusser) [inline concat.hpp:1]: Please, add a copyright and basic descriptive header.

File: concat.hpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+#include <hls_stream.h>
+#include <ap_int.h>
+#include <algorithm>
+
+
+template<typename Tsrc, typename Tdst>
+void convert_vector(Tsrc& src, Tdst& dst){
```
Comment (preusser) [inline concat.hpp:9]: If in this header file, this is a _local_ utility. Move into the anonymous namespace below.
Also, request `#pragma HLS inline`.
If only used for `hls::vector`, make this explicit in the signature and use this as an opportunity to enforce that `src` and `dst` are of the same size. Use the _inferrable_ vector length parameter as the loop bound below.

File: concat.hpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	if(src[sel].read_nb(y)) {
+		dst.write(y);
+		if(cnt >= 0)  cnt--;
+		else {
+			sel = (sel == N-1)? decltype(sel)(0) : decltype(sel)(sel+1);
+			cnt = CNT_INIT[sel];
+		}
+	}
+
+} // concat()
```
Comment (preusser) [inline concat.hpp:45]: Make consistent after name change above.

File: concat.hpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	if(reader.read_nb(sel, y, src...)) {
+		dst.write(y);
+		if(cnt >= 0)  cnt--;
+		else {
+			sel = (sel == N-1)? decltype(sel)(0) : decltype(sel)(sel+1);
+			cnt = CNT_INIT[sel];
+		}
+	}
+
+} // concat()
```
Comment (preusser) [inline concat.hpp:118]: Make consistent.

File: concat.hpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
```
Comment (preusser) [inline concat.hpp:119]: Remove trailing emtpy line.

File: split.hpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#include "utils.hpp"
```
Comment (preusser) [inline split.hpp:1]: Add license and description header.

File: split.hpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        x = src.read();
+        dst[sel].write(x);
+        if(cnt >= 0)
+            cnt--;
+        else{
+            sel = (sel == N-1) ? decltype(sel)(0) : decltype(sel)(sel+1);
+            cnt = CNT_INIT[sel]; 
+        }
+    }
+}
```
Comment (preusser) [inline split.hpp:36]: End line properly.

File: tb/concat_tb.cpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
+#include <iostream>
```
Comment (preusser) [inline tb/concat_tb.cpp:1]: Add header.

File: tb/concat_top.cpp [hunk 9]
```diff
@@ -0,0 +1,27 @@
+#include <hls_stream.h>
```
Comment (preusser) [inline tb/concat_top.cpp:1]: Add header.

File: tb/data/concat_config.h [hunk 10]
```diff
@@ -0,0 +1,22 @@
+#define SIMD 2
```
Comment (preusser) [inline tb/data/concat_config.h:1]: Use proper compile-time typing:
`constexpr unsigned  SIMD=2;`

File: tb/data/concat_config.h [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+#define NUM_INPUTS 4
+
+#define EL_TYPE0 ap_uint<3>
```
Comment (preusser) [inline tb/data/concat_config.h:4]: `using  EL_TYPE0 = ap_uint<3>;`

File: tb/data/concat_config.h [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+#define IN_TYPE3 hls::vector<EL_TYPE3, SIMD>
+#define OUT_TYPE hls::vector<EL_TYPE_OUT, SIMD>
+
+#define NUM_FOLDS0 1
+#define NUM_FOLDS1 1
+#define NUM_FOLDS2 3
+#define NUM_FOLDS3 2
+#define TOTAL_FOLDS 7
+
+#define NUM_VECTORS 3
```
Comment (preusser) [inline tb/data/concat_config.h:22]: Use something like `REP_COUNT`, `TEST_COUNT`, ...
End last line of file properly.

File: tb/data/split_config.h [hunk 13]
```diff
@@ -0,0 +1,13 @@
+#define SIMD 2
```
Comment (preusser) [inline tb/data/split_config.h:1]: See all above.

File: tb/split_tb.cpp [hunk 14]
```diff
@@ -0,0 +1,60 @@
+#include <iostream>
```
Comment (preusser) [inline tb/split_tb.cpp:1]: Add header.

File: tb/split_tb.cpp [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+					IN_TYPE exp = expected[i].read();
+					if(y != exp)
+						std::cerr << "ERROR! Unexpected output nr " << out_cnt << ". Got: " << y << " Expected: " << exp << std::endl;
+					out_cnt++;
+				}
+			}
+		}
+	}
+}
+
```
Comment (preusser) [inline tb/split_tb.cpp:59]: Remove trailing empty lines.

File: tb/split_top.cpp [hunk 16]
```diff
@@ -0,0 +1,21 @@
+#include <hls_stream.h>
```
Comment (preusser) [inline tb/split_top.cpp:1]: Add header.

**Review by preusser — COMMENTED** [review]
Thanks, @mdanilow!
Please, see the comments.

File: concat.hpp (hunk 1 above)
Comment (mdanilow) [reply concat.hpp:1]: Done

File: concat.hpp (hunk 3 above)
Comment (mdanilow) [reply concat.hpp:45]: Done

File: tb/data/concat_config.h [hunk 17]
```diff
@@ -0,0 +1,23 @@
+#include <ap_int.h>
+
+#define NUM_INPUTS 4
```
Comment (preusser) [inline tb/data/concat_config.h:3]: Sorry if I was not quite clear. There's nothing special about `SIMD`. Please, replace al of these `#define`s by a typed `constexpr`.

**Pull request merged**